File: frontend/lib/services/diary_service.dart
// lib/services/diary_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';

import 'package:family_app/api/api_client.dart';
import 'package:family_app/models/diary.dart';

/// ============================================================================
/// DiaryService
/// ----------------------------------------------------------------------------
/// - 일기 관련 API 호출 전담
/// - ID 타입은 전부 int 기준
/// ============================================================================
class DiaryService {
  // ---------------------------------------------------------------------------
  // 가족별 일기 목록 조회
  // GET /diaries/family/{family_id}
  // ---------------------------------------------------------------------------
  Future<List<Diary>> fetchDiariesByFamily(int familyId) async {
    try {
      final data = await ApiClient.get(
        '/diaries/family/$familyId',
      ) as List<dynamic>;

      return data
          .map((e) => Diary.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e, stack) {
      debugPrint('fetchDiariesByFamily 오류: $e\n$stack');
      rethrow;
    }
  }

  // ---------------------------------------------------------------------------
  // 일기 단건 조회 (사진 포함)
  // GET /diaries/{diary_id}
  // ---------------------------------------------------------------------------
  Future<Diary> fetchDiaryDetail(int diaryId) async {
    try {
      final data = await ApiClient.get(
        '/diaries/$diaryId',
      ) as Map<String, dynamic>;

      return Diary.fromJson(data);
    } catch (e, stack) {
      debugPrint('fetchDiaryDetail 오류: $e\n$stack');
      rethrow;
    }
  }

  // ---------------------------------------------------------------------------
  // 일기 생성
  // POST /diaries
  // ---------------------------------------------------------------------------
  Future<Diary> createDiary({
    required int familyId,
    required int writerId,
    required String title,
    required String content,
    required DateTime date,
    int? photoId,
  }) async {
    try {
      final data = await ApiClient.post(
        '/diaries',
        {
          'family_id': familyId,
          'writer_id': writerId,
          'title': title,
          'content': content,
          'diary_date': date.toIso8601String(),
          'photo_id': photoId,
        },
      );

      return Diary.fromJson(data as Map<String, dynamic>);
    } catch (e, stack) {
      debugPrint('createDiary 오류: $e\n$stack');
      rethrow;
    }
  }

  // ---------------------------------------------------------------------------
  // 일기 수정
  // PUT /diaries/{diary_id}
  // ---------------------------------------------------------------------------
  Future<void> updateDiary({
    required int diaryId,
    String? title,
    String? content,
    DateTime? date,
    int? photoId,
  }) async {
    try {
      await ApiClient.put(
        '/diaries/$diaryId',
        {
          'title': title,
          'content': content,
          'diary_date': date?.toIso8601String(),
          'photo_id': photoId,
        },
      );
    } catch (e, stack) {
      debugPrint('updateDiary 오류: $e\n$stack');
      rethrow;
    }
  }

  // ---------------------------------------------------------------------------
  // 일기 삭제 (soft delete)
  // DELETE /diaries/{diary_id}
  // ---------------------------------------------------------------------------
  Future<void> deleteDiary(int diaryId) async {
    try {
      await ApiClient.delete('/diaries/$diaryId');
    } catch (e, stack) {
      debugPrint('deleteDiary 오류: $e\n$stack');
      rethrow;
    }
  }
}

File: backend/app/services/diary_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete
from datetime import datetime

from app.models.diary import Diary
from app.models.photo import Photo
from app.models.album import Album
from app.models.photo_diary import PhotoDiary
from app.models.user import AppUser
from app.services.family_service import get_my_family


# -------------------------------
# 일기 생성 (+ 사진 연결)
# -------------------------------
async def create_diary(
    db: AsyncSession,
    user: AppUser,
    title: str,
    content: str,
    diary_date,
    photo_ids: list[int] | None = None,
):
    # 1️⃣ 가족 확인
    family = await get_my_family(db, user)
    if not family:
        raise Exception("Family not found")

    # 2️⃣ 일기 생성
    diary = Diary(
        family_id=family.id,
        author_id=user.id,
        title=title,
        content=content,
        diary_date=diary_date,
    )
    db.add(diary)
    await db.flush()  # diary.id 확보

    # 3️⃣ 사진 연결 (있다면)
    if photo_ids:
        result = await db.execute(
            select(Photo)
            .join(Album, Photo.album_id == Album.id)
            .where(
                Photo.id.in_(photo_ids),
                Album.family_id == family.id,
                Photo.deleted_at.is_(None),
            )
        )
        photos = result.scalars().all()

        if len(photos) != len(photo_ids):
            raise Exception("Invalid photo included")

        for photo in photos:
            db.add(
                PhotoDiary(
                    photo_id=photo.id,
                    diary_id=diary.id,
                )
            )

    await db.commit()
    await db.refresh(diary)
    return diary


# -------------------------------
# 일기 단건 조회 (내 가족)
# -------------------------------
async def get_diary(
    db: AsyncSession,
    user: AppUser,
    diary_id: int,
):
    family = await get_my_family(db, user)
    if not family:
        return None

    result = await db.execute(
        select(Diary)
        .where(
            Diary.id == diary_id,
            Diary.family_id == family.id,
            Diary.deleted_at.is_(None),
        )
    )
    return result.scalar_one_or_none()


# -------------------------------
# 내 가족 일기 목록
# -------------------------------
async def list_diaries(
    db: AsyncSession,
    user: AppUser,
):
    family = await get_my_family(db, user)
    if not family:
        return []

    result = await db.execute(
        select(Diary)
        .where(
            Diary.family_id == family.id,
            Diary.deleted_at.is_(None),
        )
        .order_by(Diary.diary_date.desc(), Diary.created_at.desc())
    )
    return result.scalars().all()


# -------------------------------
# 일기 수정 (+ 사진 재연결)
# -------------------------------
async def update_diary(
    db: AsyncSession,
    user: AppUser,
    diary_id: int,
    title: str | None = None,
    content: str | None = None,
    diary_date=None,
    photo_ids: list[int] | None = None,
):
    diary = await get_diary(db, user, diary_id)
    if not diary:
        return False

    diary.title = title or diary.title
    diary.content = content or diary.content
    diary.diary_date = diary_date or diary.diary_date
    diary.updated_at = datetime.utcnow()

    if photo_ids is not None:
        # 기존 연결 제거
        await db.execute(
            delete(PhotoDiary).where(PhotoDiary.diary_id == diary_id)
        )

        if photo_ids:
            family = await get_my_family(db, user)

            result = await db.execute(
                select(Photo)
                .join(Album, Photo.album_id == Album.id)
                .where(
                    Photo.id.in_(photo_ids),
                    Album.family_id == family.id,
                    Photo.deleted_at.is_(None),
                )
            )
            photos = result.scalars().all()

            if len(photos) != len(photo_ids):
                raise Exception("Invalid photo included")

            for photo in photos:
                db.add(
                    PhotoDiary(
                        photo_id=photo.id,
                        diary_id=diary_id,
                    )
                )

    await db.commit()
    return True


# -------------------------------
# 일기 삭제 (soft delete)
# -------------------------------
async def delete_diary(
    db: AsyncSession,
    user: AppUser,
    diary_id: int,
):
    diary = await get_diary(db, user, diary_id)
    if not diary:
        return False

    diary.deleted_at = datetime.utcnow()
    await db.commit()
    return True


# -------------------------------
# 사진 상세 + 연결된 일기 (내 가족)
# -------------------------------
async def get_photo_with_diaries(
    db: AsyncSession,
    user: AppUser,
    photo_id: int,
):
    family = await get_my_family(db, user)
    if not family:
        return None

    photo_result = await db.execute(
        select(Photo)
        .join(Album, Photo.album_id == Album.id)
        .where(
            Photo.id == photo_id,
            Album.family_id == family.id,
            Photo.deleted_at.is_(None),
        )
    )
    photo = photo_result.scalar_one_or_none()
    if not photo:
        return None

    diaries_result = await db.execute(
        select(Diary)
        .join(PhotoDiary, Diary.id == PhotoDiary.diary_id)
        .where(
            PhotoDiary.photo_id == photo_id,
            Diary.deleted_at.is_(None),
        )
    )

    return {
        "photo": photo,
        "diaries": diaries_result.scalars().all(),
    }
